Add support for Tahoma smoke sensor, light switch and two awning io components

fucm opened the pull request:
## Description:
Add support for 4 Tahoma devices:
- io:VerticalExteriorAwningIOComponent: a new cover type
- io:HorizontalAwningIOComponent: a new cover type (just added, because I have them not available for testing - they should behave like the vertical ones)
- io:OnOffLightIOComponent: light switch
- rtds:RTDSSmokeSensor: smoke sensor

Additional improvements:
- added all available state attributes, like battery warning, rssi level, memorised position, lock timer/level/originator.
- added special icon for the awnings, if they are locked by wind or something else.
- added special icon for the smoke sensor, if fire is detected or the battery is low.
- increase sensor update time from 10s to 120s

**Related issue (if applicable):** fixes #13965

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5590

## Example entry for `configuration.yaml` (if applicable):
nothing

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - No new dependencies.
  - New file already covered by `components/*/tahoma.py` (`.coveragerc`).

If the code does not interact with devices:
  - N/A (interacts with devices).

fucm committed:
Add support for 2 cover types, a smoke detector and a light switch:
- 'io:VerticalExteriorAwningIOComponent': 'cover'
- 'io:HorizontalAwningIOComponent': 'cover'
- 'io:OnOffLightIOComponent': 'switch'
- 'rtds:RTDSSmokeSensor': 'smoke'

fucm committed:
Add support for 2 cover types, a smoke detector and a light switch:
- 'io:VerticalExteriorAwningIOComponent': 'cover'
- 'io:HorizontalAwningIOComponent': 'cover'
- 'io:OnOffLightIOComponent': 'switch'
- 'rtds:RTDSSmokeSensor': 'smoke'

fucm committed:
Add support for 2 cover types, a smoke detector and a light switch:
- 'io:VerticalExteriorAwningIOComponent': 'cover'
- 'io:HorizontalAwningIOComponent': 'cover'
- 'io:OnOffLightIOComponent': 'switch'
- 'rtds:RTDSSmokeSensor': 'smoke'

fucm committed:
Add available state_attributes to all Tahoma devices

fucm committed:
Add available state_attributes to all Tahoma devices

fucm committed:
Add available state_attributes to all Tahoma devices

fucm committed:
Update Tahoma components to:
- all: provide available state attributes
- binary_sensor: smoke sensor can show battery warning icon
- sensor: increase sensor update time from 10s to 120s

fucm committed:
Update Tahoma components to:
- all: provide available state attributes
- binary_sensor: smoke sensor can show battery warning icon
- sensor: increase sensor update time from 10s to 120s

fucm committed:
Merge branch 'tahoma-io' of https://github.com/fucm/home-assistant into tahoma-io

homeassistant commented:
Hi @fucm, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#14931).

Once you do that we will be able to review and accept this pull request.

Thanks!

awarecan committed:
Upgrade python-nest, add security_state sensor, nest.set_mode service set ETA as well (#14901)

ejel committed:
Add more test cases for samsungtv (#14900)

More test cases to cover retry logic added in 58a1c3839

tchellomello committed:
Refactored Arlo component and enhanced Arlo API queries and times (#14823)

* start arlo refactoring

* Refactored Arlo Hub to avoid uncessary and duplicated GETs to Arlo API

* Refactored Arlo camera component to avoid duplicate queries

* Added debug and error messages when video is not found

* Transformed Arlo Control Panel to Sync

* Makes linter happy

* Uses total_seconds() for scan_interval

* Added callback and fixed scan_interval issue

* Disable multiple tries and supported custom modes set in Arlo

* Bump PyArlo version to 0.1.4

* Makes lint happy

* Removed ArloHub object and added some tweaks

* Fixed hub_refresh method

* Makes lint happy

* Ajusted async syntax and added callbacks decorators

* Bump PyArlo version to 0.1.6 to include some enhacements

* Refined code

awarecan committed:
Use cv.time_period instead of cv.time_period_str (#14938)

c7h committed:
Add support for KIWI Door Locks (#14485)

* initial commit for kiwi door locks

bugfixes

improved attribute display

flake8

more style adjustments

* added session handling

flake8

* added requirements_all

reordered imports and flake8

attempt to pelase a very picky linter

also pleasing pylint now :)

* re-try the build

* added kiwi.py to .coveragerc

* reorganized datetime handling and attribute naming

* created pypi package for door lock library

* updated requirements_all.txt

* code review changes

* added async lock state reset for locking state

* refactored lat/lon attribute updates

* initial locked state changed from undefined to locked

* refactored is_locked property check

* handling authentication exception in setup_platform

* added more check in setup_platform

* code review changes: return type in setup_platform

* fixed logging issue

* event handling in main thread

* updated kiwiki-client to version 0.1.1

* renamed alias e to exc

jedi7 committed:
Fix snapcast uuid to be more unique (#14925)

Current uuid is ok when using only 1 snapserver
New uuid is needed when using multiple snapserver

Because the client can connect to more snapservers and
then uuid based on client MAC is not enough

balloob commented on homeassistant/components/cover/tahoma.py:
```diff
@@ -27,27 +36,84 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
-        except KeyError:
-            return None
+            if self._position <= 5:
+                self._position = 0
```
Why is this?

fucm replied:
Actually, I don't know. It was there since the beginning (#10652). So, I kept the code.

balloob commented on homeassistant/components/cover/tahoma.py:
```diff
@@ -66,13 +131,56 @@ def device_class(self):
         return None
 
+    @property
+    def state_attributes(self):
```
Do not override `state_attributes`. Devices can override `device_state_attributes`

fucm replied:
Ok, I'll update it.

balloob commented on homeassistant/components/cover/tahoma.py:
```diff
@@ -14,6 +14,15 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+ATTR_CLOSURE = 'closure'
```
Please limit PRs to 1 feature per PR next time. So changes to binary sensor and cover: 2 prs.

fucm replied:
I will do so in the future. Sorry.

balloob commented on homeassistant/components/switch/tahoma.py:
```diff
@@ -48,4 +92,33 @@ def toggle(self, **kwargs):
+        return bool(self._state == STATE_ON)
+
+    @property
+    def state_attributes(self):
```
device_state_attributes

fucm replied:
Ok

balloob commented on homeassistant/components/switch/tahoma.py:
```diff
@@ -48,4 +92,33 @@ def toggle(self, **kwargs):
+                self.tahoma_device.active_states['core:StatusState']
+        if 'core:PriorityLockTimerState' in self.tahoma_device.active_states:
+            attr[ATTR_LOCK_TIMER] = \
+                self.tahoma_device.active_states['core:PriorityLockTimerState']
```
Is this the same info as the cover?

fucm replied:
Yes, it is the same type of info, but from the switch. Actually I do not know, for what the PriorityLockTimerState is used in case of a switch. For example a cover is blocked by a wind sensor for a certain time, in case of strong wind.

fucm commented on homeassistant/components/cover/tahoma.py:
```diff
@@ -66,13 +131,56 @@ def device_class(self):
+            attr[ATTR_STATUS] = self.tahoma_device.active_states[
+                'core:StatusState']
+        if self._lock_timer is not None:
+            attr[ATTR_LOCK_TIMER] = self._lock_timer
```
I read the docs about the states again and I don't understand, why the lock timer should go into the state machine. It's an attribute of the device for how long it is locked (before it can be used again, e.g. open cover after wind has been detected). It relates directly to lock_level and lock_originator. Could you please point me to an example or the appropriate documentation?

balloob replied:
You can store timestamps, so store the exact time when it can be used. Don't store relative times that change on every update.

fucm committed:
Use device_state_attributes instead of state_attributes in the platform code

fucm committed:
Merge branch 'tahoma-io' of https://github.com/fucm/home-assistant into tahoma-io

fucm commented:
@balloob is anything open from your review? I don't plan any other changes.

balloob commented:
It seems like you have some commits in this branch that are not yours, please clean it up.

balloob commented:
https://xkcd.com/1597/

fucm commented:
I will create smaller pull requests for each component. Closing this pull request.
